src/wayland/buffer.rs

```diff
@@ -150,6 +167,14 @@ impl WaylandBuffer {
     pub fn released(&self) -> bool {
         self.released.load(Ordering::SeqCst)
     }
+
+    fn len(&self) -> usize {
+        self.width as usize * self.height as usize
+    }
+
+    pub unsafe fn mapped_mut(&mut self) -> &mut [u32] {
+        unsafe { slice::from_raw_parts_mut(self.map.as_mut_ptr() as *mut u32, self.len()) }
+    }
```

Comment:
This should be guarded by and error out, if the buffer was not previously released.

Note, that on some wayland compositors you don't even need a back-buffer, as the compositor will release any submitted memory buffer as soon as it has uploaded the buffer as a texture. Others will hold on to it and there is no guarantee (although all compositors do that in practice), that two buffers are enough.

In practice the client *should* keep a stack of buffers, creating new ones, if all are still hold by the compositor and re-using released ones instead, if possible.

Reply:
There's already code in `wayland/mod.rs` that waits for buffer release, so that should be handled already. Though there should be more comments about safety here.

> Note, that on some wayland compositors you don't even need a back-buffer

But there's no way the compositor explicitly communicates this, right? So we'll still end up allocating at least a second buffer if the application code wants to write to a buffer before it's been released. Which is likely if we can't guarantee that only happens on resize/frame callback/etc.

So as currently implemented this always uses two buffers for a surface, and blocks if the back buffer isn't released yet. I don't know if anything more complicated is typical or likely to be worthwhile.

Reply:
> But there's no way the compositor explicitly communicates this, right?

Nope, but it also doesn't communicate, if it needs more than two. If you get a frame_callback, you should allocate a new buffer, if none are free. That is when you will know.
Technically blocking indefinitely is not against the spec, but allocating at max two buffers could render the app unresponsive on a valid compositor with an unusual implementation.

Reply:
Allowing allocating an unlimited number of buffers can cause an application submitting buffers too fast to crash. And it caused Gnome Shell to freeze temporarily: https://github.com/rust-windowing/softbuffer/issues/48

Looking at `mesa/src/egl/drivers/dri2/egl_dri2.h` and `mesa/src/egl/drivers/dri2/platform_wayland.c`, it seems it uses an array of 4 buffers? Then it blocks, though it uses a roundtrip:

```c
/* If we don't have a buffer, then block on the server to release one for
 * us, and try again. wl_display_dispatch_queue will process any pending
 * events, however not all servers flush on issuing a buffer release
 * event. So, we spam the server with roundtrips as they always cause a
 * client flush.
 */
if (wl_display_roundtrip_queue(dri2_dpy->wl_dpy,
                               dri2_surf->wl_queue) < 0)
    return -1;
```

Reply:
Yeah, by default `EglSwapBuffers` blocks on a frame callback, unless you change `EglSwapInterval` to `0`. But that blocks indefinitely for windows that are occluded, so that's not great either.

The correct solution for the animation example is https://github.com/rust-windowing/winit/issues/2412.